gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/oneconvergence_service_node_driver.py

```diff
@@ -467,47 +469,187 @@ def update_policy_target_removed(self, context, policy_target):
+                    'get_resource': fw_rule_name})
+            i += 1
+
+    @log.log
+    def update_firewall_template(self, context, stack_template,
+                                 ptg_to_not_configure=None):
+        _exist, consumer_ptgs, prov_unset = self.check_for_existing_service(
+            context)
+
+        # if not _exist:
+        #     return
+        filters = {'id': consumer_ptgs}
+        consumer_ptgs_details = context._gbp_plugin.get_policy_target_groups(
+            context.plugin_context, filters)
+
+        provider_cidr = context.core_plugin._get_subnet(
+            context._plugin_context, context.provider['subnets'][0])['cidr']
+
+        is_template_aws_version = stack_template.get(
+                                        'AWSTemplateFormatVersion', False)
+        resources_key = 'Resources' if is_template_aws_version else 'resources'
+        properties_key = ('Properties' if is_template_aws_version
+                          else 'properties')
+        fw_rule_keys = self._get_all_heat_resource_keys(
+            stack_template[resources_key], is_template_aws_version,
+            'OS::Neutron::FirewallRule')
+        fw_policy_key = self._get_all_heat_resource_keys(
+            stack_template['resources'], is_template_aws_version,
+            'OS::Neutron::FirewallPolicy')[0]
+        fw_template_properties = dict(
+            resources_key=resources_key, properties_key=properties_key,
+            is_template_aws_version=is_template_aws_version,
+            fw_rule_keys=fw_rule_keys,
+            fw_policy_key=fw_policy_key)
+
+        for consumer in consumer_ptgs_details:
+            if consumer['id'] == ptg_to_not_configure:
+                continue
+            fw_template_properties.update({'name': consumer['id'][:3]})
+            consumer_cidr = context.core_plugin._get_subnet(
+                context.plugin_context, consumer['subnets'][0])['cidr']
+            self.append_firewall_rule(context, stack_template, provider_cidr,
+                                      consumer_cidr, fw_template_properties)
+        return stack_template
+
+    @log.log
     def create(self, context):
         heatclient = self._get_heat_client(context.plugin_context)
 
-        stack_template, stack_params = self._fetch_template_and_params(context)
-
         stack_name = ("stack_" + context.instance['name'] +
                       context.current_node['name'] +
                       context.instance['id'][:8] +
                       context.current_node['id'][:8])
         # Heat does not accept space in stack name
         stack_name = stack_name.replace(" ", "")
+        stack_template, stack_params = self._fetch_template_and_params(
+            context)
+
+        service_type = context.current_profile["service_type"]
+        if (service_type == pconst.LOADBALANCER or
+                (service_type == pconst.FIREWALL and
+                 not context.is_consumer_external)
+            ):
```

Comment:
Looks fine. However, we should open a new task/bug to handle the case of when the multiple consumers are : External Policy and Internal PTG